clients/drcachesim/tracer/syscall_pt_trace.cpp

```diff
@@ -260,19 +265,17 @@ syscall_pt_trace_t::trace_data_dump(drpttracer_output_autoclean_t &output)
     pdb_header[PDB_HEADER_NUM_ARGS_IDX].syscall_args_num.args_num = 0;
 
     /* Initialize the size of the PDB data. */
-    pdb_header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_data_boundary.data_size =
-        SYSCALL_METADATA_SIZE +
-        pdb_header[PDB_HEADER_NUM_ARGS_IDX].syscall_args_num.args_num * sizeof(uint64_t) +
-        data->pt_size;
-
-    /* Write the buffer header to the output file */
-    if (write_file_func_(output_file_, pdb_header, PT_DATA_PDB_HEADER_SIZE) == 0) {
-        ASSERT(false, "Failed to write the trace data's header to the output file");
-        return false;
-    }
+    pdb_header[PDB_HEADER_DATA_BOUNDARY_IDX].pt_data_boundary.data_size = data->pt_size +
+        PT_DATA_PDB_HEADER_SIZE +
+        pdb_header[PDB_HEADER_NUM_ARGS_IDX].syscall_args_num.args_num * sizeof(uint64_t);
 
-    /* Write the syscall's PT data to the output file */
-    if (write_file_func_(output_file_, data->pt_buffer, data->pt_size) == 0) {
+    /* We want to write the data in a single call to write_file_func_ for easier 3P
+     * integration.
+     * XXX: Could we avoid the second memcpy which may be potentially large?
```

Comment:
Is this better in the end?  As you say you have an extra copy -- doesn't that outweigh the overhead of a 2nd write call?  Why is a 2nd write call so bad?  At first glance the original looks better to me.

Reply:
We're not trying to avoid a 2nd write call for performance reasons. See other comment.

Reply:
I refactored 3p code to work with separate writes for header and data.